dvc/remote/local/__init__.py

```diff
@@ -573,3 +573,51 @@ def _get_unpacked_dir_names(self, checksums):
             if self.is_dir_checksum(c):
                 unpacked.add(c + self.UNPACKED_DIR_SUFFIX)
         return unpacked
+
+    def _needs_checkout(self, path_info, checksum):
+        # NOTE: In case if path_info is already cached, and cache type is
+        # 'copy' we would like to avoid relinking.
+        # `symlink/hardlink/reflink` is cheap, but if we need `copy` links,
+        # then to save on copying we should only relink if we already have
+        # `hardlink` or a `symlink`, we don't care about reflinks, because
+        # they are indistinguishable from copy anyway.
+
+        if (
+            not self.changed(path_info, {self.PARAM_CHECKSUM: checksum})
+            and not System.is_hardlink(path_info)
+            and not System.is_symlink(path_info)
+            and self._is_cache_copy(path_info)
+        ):
+            if self.protected:
+                self.protect(path_info)
+            else:
+                # NOTE: we can unprotect, because `hardlink/symlink` check
+                # has been performed before, so no chance of copying,
+                # only chmod-ing.
+                self.unprotect(path_info)
+
+            msg = "File '{}' didn't change"
+            logger.debug(msg.format(str(path_info)))
+            return False
+        return True
+
+    def _is_cache_copy(self, path_info):
+        # NOTE: path_info required to make test reliable, when cache is on
+        # different fs than path_info
+
+        if self.cache_types[0] == "copy":
```

Comment:
Also, if we have this shortcurcuiting in place and we behave this way already in link, then there is no reason to leave cache file hanging, as we could just cache the return value of _is_cache_copy?

Reply:
True, I will refactor

Reply:
We don't need to cache return value, it is already naturally cached by this short-circuiting.

Caching method result has no sense since it has a side effect - reaping `self.cache_types`.

Reply:
@Suor True, but there is no need to leave the cache file hanging.

Reply:
I was not talking about files here.